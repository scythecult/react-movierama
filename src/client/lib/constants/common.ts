export const STORED_MOVIE_KEY = 'movierama_saved_movie';

export const STORED_SEATS_KEY = 'movierama_saved_seats';

export const DEFAULT_CANVAS_WIDTH = 400;

export const DEFAULT_CANVAS_HEIGHT = 400;

export const DEFAULT_TICKET_TYPE_ID = 1;

export const SeatTypeMap = {
  REGULAR: 1,
  COMFORT: 2,
  VIP: 3,
  WHEELCHAIR: 4,
} as const;

export const SeatStateMap = {
  FREE: 1,
  OCCUPIED: 2,
  SELECTED: 3,
} as const;

export const SEAT_TYPE: Record<string, string> = {
  [SeatTypeMap.REGULAR]: 'regular',
  [SeatTypeMap.COMFORT]: 'comfort',
  [SeatTypeMap.VIP]: 'vip',
  [SeatTypeMap.WHEELCHAIR]: 'wheelchair',
} as const;

export const SEAT_STATE: Record<string, string> = {
  [SeatStateMap.FREE]: 'free',
  [SeatStateMap.OCCUPIED]: 'occupied',
  [SeatStateMap.SELECTED]: 'selected',
} as const;
