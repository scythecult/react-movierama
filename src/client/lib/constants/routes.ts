export const AppRoute = {
  ROOT: '/',
  ORDER_PAGE: '/order',
};
