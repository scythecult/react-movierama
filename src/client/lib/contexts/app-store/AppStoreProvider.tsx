import type { PropsWithChildren } from 'react';
import { useState } from 'react';
import { type CreateAppStore, createAppStore } from '../../store/appStore';
import { AppStoreContext } from './AppStoreContext';

export const AppStoreProvider = (props: PropsWithChildren) => {
  const [appStore] = useState<CreateAppStore | null>(() => createAppStore);
  const { children } = props;

  return <AppStoreContext value={appStore}>{children}</AppStoreContext>;
};
