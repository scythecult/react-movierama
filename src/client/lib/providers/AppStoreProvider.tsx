import type { PropsWithChildren } from 'react';
import { useRef } from 'react';
import { AppStoreContext } from '../contexts/AppStoreContext';
import { type CreateAppStore, createAppStore } from '../store/appStore';

export const AppStoreProvider = (props: PropsWithChildren) => {
  const { children } = props;
  const appStoreRef = useRef<CreateAppStore | null>(null);

  if (!appStoreRef.current) {
    appStoreRef.current = createAppStore;
  }

  return <AppStoreContext.Provider value={appStoreRef.current}>{children}</AppStoreContext.Provider>;
};
