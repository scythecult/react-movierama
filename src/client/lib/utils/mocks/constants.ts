// TODO Should move to server
export const SEAT_WIDTH = 40;

export const SEAT_HEIGHT = 34;

export const SEAT_OFFSET = 10;

export const HALL_PLAN_MATRIX =
  // HALL
  [
    // ROW 1 - seat 0 - empty space width === 1 seat
    [1, 1, 0, 1, 4, 4, 1, 0, 1, 1, 1, 1, 1],
    // ROW
    [2, 2, 0, 1, 1, 1, 1, 0, 1],
    // ROW
    [1, 1, 0, 2, 2, 2, 2, 0, 1],
    // ROW
    [1, 1, 0, 3, 3, 3, 3, 0, 1],
  ];
