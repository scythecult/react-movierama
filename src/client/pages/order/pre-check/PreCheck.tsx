import { type ChangeEvent } from 'react';
import { Button } from '@/client/lib/components/button/Button';
import type { SelectedSeat } from '@/client/lib/types/OrderData';
import styles from './styles.module.css';

export type PreCheckProps = {
  selectedSeats: SelectedSeat[];
  onTicketTypeChange: (payload: SelectedSeat) => void;
  onRemoveSeatClick: (payload: SelectedSeat) => void;
}

export const PreCheck = (props: PreCheckProps) => {
  const { selectedSeats, onRemoveSeatClick, onTicketTypeChange } = props;

  if (selectedSeats.length === 0) {
    return null;
  }

  const selectedSeatNodes = selectedSeats.map((seat) => {
      const { id, place, type, seatType, price, ticketTypeId } = seat;
      const { ticketTypes } = seatType;

      return <li className={styles.preCheckItem} key={id}>
        <p>Place: {place}</p>
        <p>{type}</p>

        <div className={styles.preCheckItemControls}>
          <select className={styles.preCheckItemSelect} name="ticket-type" defaultValue={ticketTypeId}
                  onChange={(evt: ChangeEvent<HTMLSelectElement>) => {
                    const ticketTypeId = Number(evt.target.value);

                    onTicketTypeChange({ ...seat, ticketTypeId });
                  }}>
            {ticketTypes.map((ticketType) => {
                const { id, name } = ticketType;

                return <option key={id} value={id}>{name}</option>;
              },
            )}

          </select>
          <span>{price} RUB</span>
          <Button className={styles.preCheckItemRemoveButton}
                  onClick={() => onRemoveSeatClick({ ...seat, isSelected: !seat.isSelected })}>Remove
          </Button>
        </div>
      </li>;
    },
  );

  return <section className={styles.preCheck}>
    <h2 className={styles.preCheckTitle}>Chosen places</h2>
    <ul className={styles.preCheckList}>
      {selectedSeatNodes}
    </ul>
  </section>;
};