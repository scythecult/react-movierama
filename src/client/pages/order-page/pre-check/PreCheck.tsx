import type { ChangeEvent } from 'react';
import { Button } from '@/client/lib/components/button/Button';
import { SEAT_TYPE } from '@/client/lib/constants/common';
import type { CartItem } from '@/client/lib/types/OrderPageData';
import styles from './styles.module.css';

export type PreCheckProps = {
  cart: CartItem[];
  onTicketTypeChange: (seatId: number, ticketTypeId: number) => void;
  onRemoveSeatClick: (seatId: number, ticketTypeId: number) => void;
};

export const PreCheck = (props: PreCheckProps) => {
  const { cart, onRemoveSeatClick, onTicketTypeChange } = props;

  if (cart.length === 0) {
    return null;
  }

  const selectedSeatNodes = cart.map((cartItem) => {
    const { id, place, row, type, seatType, price, ticketTypeId } = cartItem;
    const { ticketTypes = [] } = seatType;
    const seatTypeName = SEAT_TYPE[type];

    return (
      <li className={styles.preCheckItem} key={id} data-test-id="pre-check-item">
        <p data-test-id="pre-check-item-place">
          {row} row, {place} place
        </p>
        <p data-test-id="pre-check-item-seat-type">{seatTypeName}</p>

        <div className={styles.preCheckItemControls} data-test-id="pre-check-item-controls">
          <select
            className={styles.preCheckItemSelect}
            name="ticket-type"
            defaultValue={ticketTypeId}
            onChange={(evt: ChangeEvent<HTMLSelectElement>) => {
              const ticketTypeId = Number(evt.target.value);

              onTicketTypeChange(id, ticketTypeId);
            }}
            data-test-id="pre-check-item-select"
            data-test-value={ticketTypeId}
          >
            {ticketTypes.map((ticketType) => {
              const { id, name } = ticketType;

              return (
                <option key={id} value={id} data-test-id="pre-check-item-select-option" data-test-value={name}>
                  {name}
                </option>
              );
            })}
          </select>
          <span data-test-id="pre-check-item-price">{price} RUB</span>
          <Button
            className={styles.preCheckItemRemoveButton}
            onClick={() => onRemoveSeatClick(id, ticketTypeId)}
            data-test-id="pre-check-remove-button"
          >
            Remove
          </Button>
        </div>
      </li>
    );
  });

  return (
    <section className={styles.preCheck} data-test-id="pre-check">
      <h2 className={styles.preCheckTitle} data-test-id="pre-check-title">
        Chosen places
      </h2>
      <ul className={styles.preCheckList} data-test-id="pre-check-list">
        {selectedSeatNodes}
      </ul>
    </section>
  );
};
