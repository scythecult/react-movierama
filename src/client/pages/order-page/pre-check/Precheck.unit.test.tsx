import { render } from '@testing-library/react';
import { describe } from 'vitest';
import { AppStoreProvider } from '@/client/lib/providers/AppStoreProvider';
import { CART_ITEMS } from '../../../../tests/constants';
import { PreCheck, type PreCheckProps } from './PreCheck';

const DEFAULT_PROPS: PreCheckProps = {
  cart: [],
  onRemoveSeatClick: () => {},
  onTicketTypeChange: () => {},
};

const buildWrappedComponent = (props = DEFAULT_PROPS) => (
  <AppStoreProvider>
    <PreCheck {...props} />
  </AppStoreProvider>
);

describe('PreCheck', () => {
  test('should correspond default layout', () => {
    const result = render(buildWrappedComponent());

    expect(result.container).toMatchSnapshot();
  });

  test('should support the "cart" prop', () => {
    let result = render(buildWrappedComponent({ ...DEFAULT_PROPS }));

    expect(result.container).toMatchSnapshot();

    result = render(buildWrappedComponent({ ...DEFAULT_PROPS, cart: CART_ITEMS }));

    expect(result.container).toMatchSnapshot();
  });
});
