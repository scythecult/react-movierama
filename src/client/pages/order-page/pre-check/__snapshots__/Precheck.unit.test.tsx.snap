// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`PreCheck > should correspond default layout 1`] = `<div />`;

exports[`PreCheck > should support the "cart" prop 1`] = `<div />`;

exports[`PreCheck > should support the "cart" prop 2`] = `
<div>
  <section
    class="_preCheck_9484b4"
  >
    <h2
      class="_preCheckTitle_9484b4"
    >
      Chosen places
    </h2>
    <ul
      class="_preCheckList_9484b4"
    >
      <li
        class="_preCheckItem_9484b4"
      >
        <p>
          1
           row, 
          1
           place
        </p>
        <p>
          regular
        </p>
        <div
          class="_preCheckItemControls_9484b4"
        >
          <select
            class="_preCheckItemSelect_9484b4"
            name="ticket-type"
          >
            <option
              value="1"
            >
              student
            </option>
            <option
              value="2"
            >
              adult
            </option>
          </select>
          <span>
            100
             RUB
          </span>
          <button
            class="_preCheckItemRemoveButton_9484b4 _button_d1d523"
          >
            Remove
          </button>
        </div>
      </li>
    </ul>
  </section>
</div>
`;
